sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java

```diff
@@ -1014,19 +1100,49 @@ private TextWriteOperation(TextSink<T> sink, Coder<T> coder) {
     private static class TextWriter<T> extends FileBasedWriter<T> {
       private static final byte[] NEWLINE = "\n".getBytes(StandardCharsets.UTF_8);
       private final Coder<T> coder;
+      private final String header;
+      private final String footer;
       private OutputStream out;
 
       public TextWriter(FileBasedWriteOperation<T> writeOperation, Coder<T> coder) {
+        this(writeOperation, coder, null, null);
+      }
+
+      public TextWriter(FileBasedWriteOperation<T> writeOperation, Coder<T> coder, String header) {
+        this(writeOperation, coder, header, null);
+      }
+
+      public TextWriter(FileBasedWriteOperation<T> writeOperation, Coder<T> coder, String header,
+                        String footer) {
         super(writeOperation);
+        this.header = header;
+        this.footer = footer;
         this.mimeType = MimeTypes.TEXT;
         this.coder = coder;
       }
 
+      private void writeLine(String line) throws IOException {
```

Comment:
`writeLine` seems a strange name for this function -- why do we think the `String` is only a line?

Javadoc should specify that `null` values will be ignored, but if the value is non-`null` then a newline will be added and the line will not be escaped.